### PR TITLE
Target Hints: Add missing param sanitization

### PR DESCRIPTION
**Pull request by swissspidy**
## What?

This fixes a serious bug in the `Gutenberg_REST_Server` class which is intended for 6.7 merge. This bug causes fatal errors for plugins like Web Stories.

#64504 updated `WP_REST_Sever` to add the `targetHints` property. Like `WP_REST_Server::dispatch()` it uses `match_request_to_handler()` to get the correct handler and then "sends" the correct headers.

The problem is that it does not sanitize the request params it gathered from `WP_REST_Request::from_url()` and `match_request_to_handler()`. So if the controller in question expects an `int`, it will get a `string` and if it has strict typing, that leads to fatal errors.

**Note:** This might warrant a 19.2.1 release.

## Why?

Bugs are bad!

## How?

Fixes the bug by validating and sanitizing the params.

I don't have the capacity right now to add tests, so if someone else can do it, that would be great.

## Testing Instructions

1. Install the Web Stories plugin
2. Open the Web Stories editor
3. See (or not see) a fatal error

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

N/A

### CODE REVIEW

**Comment by swissspidy** [comment]
Adding the "No Core Sync Required" label solely because the [core PR](https://github.com/WordPress/wordpress-develop/pull/7139) wasn't actually merged yet

File: lib/compat/wordpress-6.7/class-gutenberg-rest-server.php [hunk 1]
```diff
@@ -95,16 +95,22 @@ public static function get_response_links( $response ) {
...
-					$response->set_matched_handler( $match[1] );
-					$headers = rest_send_allow_header( $response, $server, $request )->get_headers();
-
-					foreach ( $headers as $name => $value ) {
-						$name                               = WP_REST_Request::canonicalize_header_name( $name );
-						$attributes['targetHints'][ $name ] = array_map( 'trim', explode( ',', $value ) );
+				$matched = $server->match_request_to_handler( $request );
+				if ( ! is_wp_error( $matched ) ) {
+					if ( ! is_wp_error( $request->has_valid_params() ) ) {
+						if ( ! is_wp_error( $request->sanitize_params() ) ) {
```
Comment (Mamaduka) [inline lib/compat/wordpress-6.7/class-gutenberg-rest-server.php]: Nit: Can we reduce nesting here?

Comment (swissspidy) [reply lib/compat/wordpress-6.7/class-gutenberg-rest-server.php]: Sure.

Although I don't particularly like the `$data[ $rel ][] = $attributes; continue;` repetition :-)

Comment (Mamaduka) [reply lib/compat/wordpress-6.7/class-gutenberg-rest-server.php]: Me neither, but three levels of `if` statements also feels wrong 😄

**Review by Mamaduka — APPROVED** [review]
Can confirm that this fixes the fatal error with the Web Stories plugin.

The unit test coverage would be nice to have, but it is probably not a blocker for the Gutenberg plugin fix if we want to ship a minor release as soon as possible.

**Comment by peterwilsoncc** [comment]
@swissspidy Are you able to add a link to this PR to https://github.com/WordPress/gutenberg/blob/trunk/backport-changelog/6.7/7139.md

**Comment by swissspidy** [comment]
Weird workflow, clearly I am not accustomed to it. Done now.

**Comment by swissspidy** [comment]
E2E failures are unrelated and happen because of a core change.

**Comment by TimothyBJacobs** [comment]
I'm adding tests in https://github.com/WordPress/wordpress-develop/pull/7139, is that properly testing what you were running into @swissspidy?

**Comment by swissspidy** [comment]
I think so, yes, thanks @TimothyBJacobs!

**Comment by swissspidy** [comment]
@Mamaduka do you think this warrants a minor release? if not, when is 19.3 happening?

**Pull request merged**